packages/ts/react-crud/src/autogrid.tsx

```diff
@@ -156,21 +257,20 @@ function createDataProvider<TItem>(
       pageSize,
       sort,
     };
-
     const items = await service.list(req, filter.current);
-    let size;
-    if (items.length === pageSize) {
-      size = (pageNumber + 1) * pageSize + 1;
-
-      const cacheSize = (grid as GridElementWithInternalAPI<TItem>)._dataProviderController.rootCache.size;
-      if (cacheSize !== undefined && size < cacheSize) {
-        // Only allow size to grow here to avoid shrinking the size when scrolled down and sorting
-        size = undefined;
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if (!(service as any).count) {
+      if (itemCountHolder.totalCount ?? itemCountHolder.filteredCount) {
+        console.error(
+          '"totalCount" and/or "filteredCount" props require the provided service to implement the CountService interface.',
+        );
       }
+      initDataProviderUsingInfiniteScrolling(items, pageSize, pageNumber, grid, callback);
```

Comment:
Reading the code it is confusing that there is a `initDataProviderUsingCountService` but it doesn't actually init or create anything but instead just contains some logic of the data provider request. 

Maybe it would be clearer to structure it so that there are two different data provider implementations that reuse some common parts like querying the list service and applying the `recalculateColumnWidths` timeout fix. Like, have `createInfiniteDataProvider` and `createFiniteDataProvider` functions that return a data provider with the respective implementation.